### ISSUE TITLE
Feature Request: Multi-Load - Read .idx (index file) for TAP files

### ISSUE THREAD

**Issue by Maverick-Shark** [mist-devel/c64#13]
Hi,

I would like to load tap preserved files from cassettes with Turbo Load. In a .tap preserved file (from a cassette), you have all the programs saved one per one (all of them are recorded after each other).

Then, I would like to put the virtual position of each program (as the datassette counter) to run Turbo Load program at the beginning and jump to the correct position to load the program/game. I think that it would be add in the config the option to auto-pause the loaded tape when it finish, and play again when you put load or run.

Another example, 007 A View To a Kill preserved tap game (it's only one file). It has different missions and you need to reset the machine (c64) after finish each one to load the following.
It would be good to move to the correct position.

We have different tools like TAP Manager (v.1.6.0 or later) or IDXGen v1.0, that you can download from http://www.manosoft.it/ that generates an index file (.idx). You can choose between to type of index files (Princess or Ultimate 64):
 
![image](https://github.com/mist-devel/c64/assets/10196716/0f68d088-ca7f-489a-8896-66de3f9cb58d)

I think that it will be a very good feature and useful for multi load and compilation preserved tapes.

Luigi Di Fraia's wrote this:

https://luigidifraia.wordpress.com/2020/10/13/the-origin-of-the-idx-format-supported-by-dc2n5-lc-ultimate-64-etc/

Thanks for your big job, it's awesome!!
Shark

**Comment by gyurco** [comment]
It would be mostly a firmware addition to parse the index and present an UI to choose the tape position...maybe it'll happen one day.

**Comment by Maverick-Shark** [comment]
Yes, maybe!! (I'm crossing my fingers that it will be soon). ;-P
And this tap player would de implemented on the arm/mcu, of course.

Thanks,
Shark

**Comment by Maverick-Shark** [comment]
Hi,

Could you please see Gideon code?

https://github.com/GideonZ/1541ultimate/blob/master/software/filetypes/filetype_tap.cc

Gideon has implemented it on his Ultimate64 cartridge.

![idx](https://github.com/mist-devel/c64/assets/10196716/b0dcdbb1-063e-4584-ad3b-2d54ca98d5e0)


Regards,
Sahrk

**Comment by Maverick-Shark** [comment]
Hi Gyurco,

New update, regarding Gideon code:

![image](https://github.com/user-attachments/assets/99d60121-1306-42d8-b8cc-4df1f27d4453)

https://github.com/user-attachments/assets/f5e64ca8-0f04-408e-a0df-f94d40853fa5

New c64 firmware/framework with this feature would be awesome!!

Thanks and best regards,
Shark

**Comment by Maverick-Shark** [comment]
Hi Gyurco,

Sorry to bother you, again!!! 

I've a new idea. Reviewing Gideon's code, sure that it could be better to implement it on MiST firmware (because it's in C). What do you think about it?

Regards,
Shark

**Comment by gyurco** [comment]
Do you have some example IDX/TAP files?

**Comment by Maverick-Shark** [comment]
Hi Gyurco,

Yes, you can find a lot of examples of games that need .idx files at archive.org (https://archive.org/details/ultimatetapearchive), i.e.:

[A View to a Kill - The Computer Game (1985)(Domark)](https://archive.org/download/uta_A_View_to_a_Kill-_The_Computer_Game_1985_Domark_3365)
https://archive.org/download/uta_A_View_to_a_Kill-_The_Computer_Game_1985_Domark_3365

[007 A View To A Kill (1985) (Domark).zip](https://github.com/user-attachments/files/20973766/007.A.View.To.A.Kill.1985.Domark.zip)

````
Name:                                   |   Last modified:  |   Size:
A_View_to_a_Kill-_The_Computer_Game.idx | 26-Aug-2020 18:24 | 170.0B
A_View_to_a_Kill-_The_Computer_Game.tap | 26-Aug-2020 18:25 | 1.9M
````

A_View_to_a_Kill-_The_Computer_Game.idx:
````
; Index file by The Ultimate Tape Archive Team

0x00000014   Opening
0x0004b74d   Paris Chase
0x000c3be8   City Hall
0x0012f51d   The Mine - Silicon Valley
0x001a3687   End
````

.idx is a text file with the start position of the program. Sure that you know all .tap files start with 0x14 (c64 tap file header).

![Image](https://github.com/user-attachments/assets/1b9caf29-6cb1-4031-a3d9-8e2f8a78619c)

Thanks,
Shark

**Comment by gyurco** [comment]
The IDX file in the ZIP differs:
```
01-OPENING!!,000014,04D002
02-PARIS!,04D003,0C5461
03-CITY!,0C5462,130D5A
04-MINE,130D5B,1A4E97
05-END,1A4E98,1EB580
```
The separator is a comma, and there's a length field, too. Even the field order is changed.
Seems there are multiple IDX standards?

**Comment by Maverick-Shark** [comment]
Hi Gyurco,

Yes, you are right again. Sorry!!!

They are two .idx format files, PRINCESS and Ultimate64, but at the end are practically the same (both show the hexadecimal position). Last .idx was in PRINCESS format, not for Ultimate64.

.idx Ultimate64 format is the following:

````
0x14 OPENING!!
0x4B74D PARIS!
0xC3BE8 CITY!
0x12F51D MINE
0x1A3687 END
````

But at the end, you can choose your own .idx file type or the generated by The Ultimate Tape Archive Team at Ultimate Tape Archive v5 collection:

https://ia601302.us.archive.org/view_archive.php?archive=/14/items/Ultimate_Tape_Archive_V5/Ultimate_Tape_Archive_V5.zip

That it's supported by DC2N, NOVA64 and the Ultimate products. Probably this one is the most compatible format:

````
; Index file by The Ultimate Tape Archive Team

0x00000014 Opening
0x0004b74d Paris Chase
0x000c3be8 City Hall
0x0012f51d The Mine - Silicon Valley
0x001a3687 End
````

I shared to you the IDXGen software to generate both types of index files (the .zip file contains too the .idx in Ultimate64 format).

The most important thing is, at the end, the possibility to load from an .idx file the hexadecimal position where the program starts.

[IDXGen.zip](https://github.com/user-attachments/files/21005904/IDXGen.zip)


Rgrds,
Shark

**Comment by gyurco** [comment]
I added a premilinary handler to the firmware:
https://github.com/mist-devel/mist-firmware/commit/3f02dc5c96c44a6ea127e910d5c8cba716777dc0
Need a little change in the core, too.
Then maybe I'll investigate T64 support. AFAIK it's just a bunch of .PRG files.

**Comment by Maverick-Shark** [comment]
> I added a premilinary handler to the firmware: [mist-devel/mist-firmware@3f02dc5](https://github.com/mist-devel/mist-firmware/commit/3f02dc5c96c44a6ea127e910d5c8cba716777dc0) Need a little change in the core, too. Then maybe I'll investigate T64 support. AFAIK it's just a bunch of .PRG files.

Hi,

Thanks a lot again!! Awesome work!!!

I'll wait until the firmware is compiled to update it.

Thanks and best regards,
Shark

**Comment by gyurco** [comment]
Here's a binary for MiST/SiDi/Poseidon (AT91SAM MCU), the core is updated in git.
Would be good to test before release

[firmware-250728.zip](https://github.com/user-attachments/files/21477516/firmware-250728.zip)

, I only tried with one IDX.

**Comment by Maverick-Shark** [comment]
Hi,

Tested with the new firmware and it works awesome on Poseidon board. I'm so happy!!!

Many many thanks to add this new feature. I know it has been a hard work and sure that you have dedicated a lot of time.

You said to improve it adding t64 too, but in my opinion, now, this core is awesome. I mean, not sure if we need it for t64 files too.

The idea was just that, to load .tap files from a certain position and you have made it possible, like a real datassette counter does and you have made it possible. Very useful to load magazine tapes or turbo tapes. 

Thanks and Best regards,
Shark

**Comment by gyurco** [comment]
I researched further T64 support, but seems it's nearly useless. There are not many T64 files with multiple PRGs inside.
